source/core/slang-common.h

```diff
@@ -76,28 +76,43 @@ namespace Slang
     template <typename T>
     SLANG_FORCE_INLINE T* clone(IClonable* clonable) { return (T*)clonable->clone(T::getTypeGuid()); }
 
+}
 
 // TODO: Shouldn't these be SLANG_ prefixed?
 #ifdef _MSC_VER
 #define UNREACHABLE_RETURN(x)
-#define UNREACHABLE(x)
 #else
 #define UNREACHABLE_RETURN(x) return x;
-#define UNREACHABLE(x) x;
 #endif
 
-}
+//
+// Use `SLANG_ASSUME(myBoolExpression);` to inform the compiler that the condition is true.
+// Do not rely on side effects of the condition being performed.
+//
+#if defined(__cpp_assume)
+#    define SLANG_ASSUME(X) [[assume(X)]]
+#elif SLANG_GCC
+#    define SLANG_ASSUME(X) do{if(!(X)) __builtin_unreachable();} while(0)
+#elif SLANG_CLANG
+#    define SLANG_ASSUME(X) __builtin_assume(X)
+#elif SLANG_VC
+#    define SLANG_ASSUME(X) __assume(X)
+#else
+     [[noreturn]] inline void invokeUndefinedBehaviour() {}
+#    define SLANG_ASSUME(X) do{if(!(X)) invokeUndefinedBehaviour();} while(0)
+#endif
 
+//
+// Assertions abort in debug builds, but inform the compiler of true
+// assumptions in release builds
+//
 #ifdef _DEBUG
-#define SLANG_EXPECT(VALUE, MSG) if(VALUE) {} else SLANG_ASSERT_FAILURE(MSG)
-#define SLANG_ASSERT(VALUE) SLANG_EXPECT(VALUE, #VALUE)
+#define SLANG_ASSERT(VALUE) do{if(!(VALUE)) SLANG_ASSERT_FAILURE(#VALUE);} while(0)
 #else
-#define SLANG_EXPECT(VALUE, MSG) do {} while(0)
-#define SLANG_ASSERT(VALUE) do {} while(0)
+#define SLANG_ASSERT(VALUE) SLANG_ASSUME(VALUE)
 #endif
 
 #define SLANG_RELEASE_ASSERT(VALUE) if(VALUE) {} else SLANG_ASSERT_FAILURE(#VALUE)
```

Comment:
This is not using SLANG_ASSUME?

Reply:
deliberately not using ASSUME. A release assert is one we specifically want to still message/abort on, even in release builds.

Reply:
TBH, I'm inclined to just replace it with SLANG_ASSERT as it's not really clear what it's saying: "I am certain that this condition is true, HOWEVER I'm not sure enough that I'm comfortable not checking it in release builds...". If that's the case then we should either reason further about the code to make ourselves more certain that it can't happen, or we should replace these with explicit calls to `abort()` (or whatever end of days error handling function)

Reply:
We should be careful about that because we have many places that uses release_assert because the expr inside the assert is intended to run.

Reply:
Indeed, FWIW I think that clang can warn when such expressions can't be proven side effect free.